mypy/ipc.py

```diff
@@ -0,0 +1,191 @@
+        super().__init__(name)
+        if sys.platform == 'win32':
+            timeout = timeout or 1000  # we need to set a timeout
+            while True:
+                try:
+                    _winapi.WaitNamedPipe(self.name, timeout)
+                except WindowsError as e:
+                    if e.winerror == _winapi.ERROR_SEM_TIMEOUT:
+                        raise IPCException("Timed out waiting for connection.")
+                    else:
+                        raise
+                try:
+                    self.connection = _winapi.CreateFile(
+                        self.name,
+                        _winapi.GENERIC_READ | _winapi.GENERIC_WRITE,
+                        0,
+                        _winapi.NULL,
+                        _winapi.OPEN_EXISTING,
+                        0,
+                        _winapi.NULL,
+                    )
+                except WindowsError as e:
+                    if e.winerror == _winapi.ERROR_PIPE_BUSY:
+                        raise IPCException("The connection is busy.")
+                    else:
+                        raise
+                else:
+                    break
+            _winapi.SetNamedPipeHandleState(self.connection,
+                                            _winapi.PIPE_READMODE_MESSAGE,
+                                            None,
+                                            None)
+        else:
+            self.connection = socket.socket(socket.AF_UNIX)
+            self.connection.settimeout(timeout)
+            self.connection.connect(name)
+
+    def __enter__(self) -> 'IPCClient':
+        return self
+
+
+class IPCServer(IPCBase):
+
+    BUFFER_SIZE = 2**16
+
+    def __init__(self, name: str, timeout: Optional[int] = None) -> None:
+        if sys.platform == 'win32':
+            name = r'\\.\pipe\{}-{}.pipe'.format(name, os.getpid())
+        super().__init__(name)
+        if sys.platform != 'win32':
+            self.sock_directory = tempfile.mkdtemp()
+            sockfile = os.path.join(self.sock_directory, self.name)
+            self.sock = socket.socket(socket.AF_UNIX)
+            self.sock.bind(sockfile)
+            self.sock.listen(1)
+            if timeout is not None:
+                self.sock.settimeout(timeout)
+
+    def __enter__(self) -> 'IPCServer':
+        if sys.platform == 'win32':
+            # The NamedPipe needs to be recreated every time we wait for a connection
+            self.connection = _winapi.CreateNamedPipe(self.name,
+                _winapi.PIPE_ACCESS_DUPLEX | _winapi.FILE_FLAG_FIRST_PIPE_INSTANCE,
+                _winapi.PIPE_READMODE_MESSAGE | _winapi.PIPE_TYPE_MESSAGE | _winapi.PIPE_WAIT,
+                1,  # one instance
+                self.BUFFER_SIZE,
+                self.BUFFER_SIZE,
+                0,  # PIPE_WAIT, wait for a connection of the client
+                0,  # PIPE_ACCEPT_REMOTE_CLIENTS (this is the default, so I guess?)
```

Comment:
`PIPE_ACCEPT_REMOTE_CLIENTS` seems to actually be a flag for the third arg, not the last one. (And probably we should *not* accept remote clients, I would think?)

The last arg seems to be `lpSecurityAttributes`, and I'm having trouble figuring out if the default if `NULL` is passed is what we want: "The ACLs in the default security descriptor for a named pipe grant full control to the LocalSystem account, administrators, and the creator owner. They also grant read access to members of the Everyone group and the anonymous account."

(Discussed in https://docs.microsoft.com/en-us/windows/desktop/ipc/named-pipe-security-and-access-rights)

It's not completely clear to me what exactly the implications of "read access to everyone" is? Without being to write, it can't cause any typechecking, but if it can connect and hold open a connection it could hang the server?

Reply:
Yes, it could be a DOS, since the maximum clients is 1. Looks like I will need to break out some ctypes for this and `DisconnectNamedPipe`.